ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/CustodyGroupCountManagerImpl.java

```diff
@@ -93,29 +94,69 @@ public void onSlot(final UInt64 slot) {
       return;
     }
 
-    if (!updateEpoch(spec.computeEpochAtSlot(slot))) {
+    final UInt64 currentEpoch = spec.computeEpochAtSlot(slot);
+
+    final Map<UInt64, PreparedProposerInfo> preparedValidators =
+        proposersDataManager.getPreparedProposerInfo();
+
+    if (detectGenesisInitialization(currentEpoch, preparedValidators)) {
+      updateEpoch(currentEpoch);
       return;
     }
 
-    final Map<UInt64, PreparedProposerInfo> preparedProposerInfo =
-        proposersDataManager.getPreparedProposerInfo();
-    if (preparedProposerInfo.isEmpty()) {
+    if (!updateEpoch(currentEpoch)) {
+      return;
+    }
+
+    if (preparedValidators.isEmpty()) {
       updateCustodyGroupCount(initCustodyGroupCount);
       return;
     }
 
+    computeAndUpdateCustodyGroupCount(preparedValidators);
+  }
+
+  private boolean detectGenesisInitialization(
+      final UInt64 currentEpoch, final Map<UInt64, PreparedProposerInfo> preparedValidators) {
+    if (!currentEpoch.isZero()) {
```

Comment:
how about such order and little adjustment to logic, looks a bit safer for me
```java
    if (genesisInitialized) {
        return false;
        }
        
    if (!currentEpoch.isZero()) {
      genesisInitialized = true;  
      return false;
    }

    if (preparedValidators.isEmpty()) {
      return false;
    }

```

Reply:
No I actually want to run that as soon as we have a some prepared validators.
Validator preparation depends on VC implementation (we call it at slot 3 IIRC), so the idea is to constantly check that along epoch 0 and do the initialization as soon as we have some validators.